pkg/cgroups/cgroupblkio.go

```diff
@@ -106,6 +109,143 @@ func GetBlkioDir() string {
+		newBlockIO.WeightDevice = append(newBlockIO.WeightDevice, ociWDP)
+	}
+	// Reset old device weights that were missing from blockIO.WeightDevice
+	for _, ociWDP := range oldBlockIO.WeightDevice {
+		if !seenDev[devMM{ociWDP.Major, ociWDP.Minor}] {
+			newBlockIO.WeightDevice = append(newBlockIO.WeightDevice, OciWeightDeviceParameters{ociWDP.Major, ociWDP.Minor, 0})
+		}
+	}
+	newBlockIO.ThrottleReadBpsDevice = resetDevRates(oldBlockIO.ThrottleReadBpsDevice, blockIO.ThrottleReadBpsDevice)
+	newBlockIO.ThrottleWriteBpsDevice = resetDevRates(oldBlockIO.ThrottleWriteBpsDevice, blockIO.ThrottleWriteBpsDevice)
+	newBlockIO.ThrottleReadIOPSDevice = resetDevRates(oldBlockIO.ThrottleReadIOPSDevice, blockIO.ThrottleReadIOPSDevice)
+	newBlockIO.ThrottleWriteIOPSDevice = resetDevRates(oldBlockIO.ThrottleWriteIOPSDevice, blockIO.ThrottleWriteIOPSDevice)
+	errors = multierror.Append(errors, SetBlkioParameters(cgroupsDir, newBlockIO))
+	return errors.ErrorOrNil()
+}
+
+// resetDevRates adds wanted rate parameters to new and resets unwated rates
+func resetDevRates(old, wanted []OciRateDeviceParameters) []OciRateDeviceParameters {
+	new := []OciRateDeviceParameters{}
+	seenDev := map[devMM]bool{}
+	for _, rdp := range wanted {
+		new = append(new, rdp)
+		seenDev[devMM{rdp.Major, rdp.Minor}] = true
+	}
+	for _, rdp := range old {
+		if !seenDev[devMM{rdp.Major, rdp.Minor}] {
+			new = append(new, OciRateDeviceParameters{rdp.Major, rdp.Minor, 0})
+		}
+	}
+	return new
+}
+
+// GetBlkioParameters returns OCI BlockIO parameters from files in cgroups blkio controller directory.
+func GetBlkioParameters(cgroupsDir string) (OciBlockIOParameters, error) {
+	var errors *multierror.Error
+	blockIO := NewOciBlockIOParameters()
+	content, err := readFromFileInDir(cgroupsDir, blkioWeightFiles)
+	if err == nil {
+		weight, err := strconv.ParseInt(strings.TrimSuffix(content, "\n"), 10, 64)
+		if err == nil {
+			blockIO.Weight = weight
+		} else {
+			errors = multierror.Append(errors, fmt.Errorf("parsing weight from %#v failed: %w", content, err))
+		}
+	} else {
+		errors = multierror.Append(errors, err)
+	}
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioWeightDeviceFiles, &blockIO.WeightDevice, nil))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleReadBpsFiles, nil, &blockIO.ThrottleReadBpsDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleWriteBpsFiles, nil, &blockIO.ThrottleWriteBpsDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleReadIOPSFiles, nil, &blockIO.ThrottleReadIOPSDevice))
+	errors = multierror.Append(errors, readDevWeightsRates(cgroupsDir, blkioThrottleWriteIOPSFiles, nil, &blockIO.ThrottleWriteIOPSDevice))
+	return blockIO, errors.ErrorOrNil()
+}
+
+func readDevWeightsRates(baseDir string, filenames []string, outWeights *[]OciWeightDeviceParameters, outRates *[]OciRateDeviceParameters) error {
```

Comment:
This interface looks a bit awkward. How about replacing `outWeights` and `outRates` with one empty interface parameter and doing a type switch in the function.

Or, is there really need to have two separate types `OciWeightDeviceParameters` and `OciRateDeviceParameters`? At a quick glance they could be replaced with one `OciDeviceParameters`, simplifying the code quite a bit.

WDYT?

Reply:
I agree. However, I'm strictly following here structures in [OCI blockio spec](https://github.com/tianon/oci-runtime-spec/blob/master/config-linux.md#block-io) in order to keep the cgroups interface OCI-like. OCI has different structs for blkioWeightDevice and blkioThrottle*. The price is extra code complexity.

This particular interface is an alternative to my first solution that had separate `readDevWeights` and `readDevRates` functions that were practically the same, as you can imagine.

Reply:
Please mark conversation resolved if this explanation makes this code bearable in your eyes.

Now that we're not really interested on `leafWeight`, we really could get rid of separate `rate` and `weight` structures and simplify the code. But maybe that's another patch?

Reply:
What do you think about the empty interface parameter I suggested?

Or then drop `readDevWeightsRates` for good and do smth like:
```
type OciDeviceParameters interface {
        append(maj, min, val,int64)
}

type OciWeightDeviceParameter struct {
        Major  int64
        Minor  int64
        Weight int64
}

/*
* Add corresponding definitions for OciRateDeviceParameters
*/

type OciWeightDeviceParameters []OciWeightDeviceParameter

func (p *OciWeightDeviceParameters) append(maj, min, val int64) { *p = append(*p, OciWeightDeviceParameter{Major: maj, Minor: min, Weight: val} }

func readMajorMinorValues(baseDir string, filenames []string, OciDeviceParameters) error {
...
}

I think this would be a bit more go-like
```

Reply:
If we start walking down that path, then I'd rather also rename the current `OciWeightDeviceParameters` and `OciRateDeviceParameters` to `OciDeviceWeight` and `OciDeviceRate` correspondingly. Then define plural-named slices of these, `OciDeviceWeights` and `OciDeviceRates`, and have them support the interface @marquiz  suggested above. Finally, I'd drop/fold `readMajMinVals()` into `readDevWeightsRates()` and rename it to `readOciDeviceParameters()`. Something like this (sans the worst typos and thinkos, of course):

```
type OciDeviceWeight struct {
    Major  int64
    Minor  int64
    Weight int64
}

type OciDeviceRate struct {
    Major  int64
    Minor  int64
    Rate   int64
}

type OciDeviceParameters interface {
    append(maj, min, val int64)
}

type OciDeviceWeights []OciDeviceWeight
type OviDeviceRates []OciDeviceRate

func (w *OciDeviceWeights) append(maj, min, val int64) {
    *w = append(*w, OciDeviceWeight{Major: maj, Minor: min, Weight: val})
}

func (r *OciDeviceRates) append(maj, min, val int64) {
    *r = append(*r, OciDeviceRate{Major: maj, Minor: min, Rate: val})
}

func readOciDeviceParameters(baseDir string, filenames []string, params OciDeviceParameters) error {
    var errors *multierror.Error
    ... // loop of reading and parsing the files
        params.append(major, minor, value)
    ...
    return errors.ErrorOrNil()
}
```

Reply:
👍exactly, more thought out naming (and expression) than what I had

Reply:
@askervin Are there any plans to address this last review comment ? Or would you prefer not to ?

Reply:
@klihub , @marquiz , I think this was very good improvement to naming and structuring the code. I walked a bit further down the path in order to clean up `updateOrAppendDevice{Weight,Rate}` functions similarly. So the interface above got the form
```go
type OciDeviceParameters interface {
	Append(maj, min, val int64)
	Update(maj, min, val int64)
}
```

The patch is now updated and tested (old and new unit tests + system test with SIGHUP-config update).